### PR TITLE
Fixes 11334/11354 Multi line code blocks

### LINKED ISSUE #11334: Cannot make multi-line code blocks in ipython
I just did a `pip install ipython` while teaching a class today and typed an if statement and then hit enter after the first line and the code executed.

This seems like a bug.

I did a `pip install ipython==6.5.0` and hitting Enter in a code block properly showed me the next indented line to type out.

I'm on Ubuntu 18.04 running ipython within tmux, though I doubt tmux is the issue here.

### PR DESCRIPTION
**Pull request by tonyfast**
This pull request merges the recent input transformer changes from master.  It builds off of the work by @bartskowron in  #11354 to resolve #11334 .  The changes pass tests locally, including the new tests provided by @Carreau; let's see what CI says.

@takluyver [I commented out an assertion](https://github.com/ipython/ipython/blame/498f5f716d102a572cac3571e13426aa002f8454/IPython/core/inputtransformer2.py#L252).  I am not sure what the statement, is this change ok?

### CODE REVIEW

**Comment by tonyfast** [comment]
Anyone have any ideas why this may be passing on everything, but Python 3.7dev?

**Comment by Carreau** [comment]
> Anyone have any ideas why this may be passing on everything, but Python 3.7dev?

Maybe a change of API in generate_token or similar on Pyhton 3.7 dev.

**Comment by tonyfast** [comment]
I'll scope it out. Thanks.

**Comment by Carreau** [comment]
that's https://bugs.python.org/issue33899

**Comment by tonyfast** [comment]
Thanks for these references @Carreau .  I added some conditions for the newline tokens.  Let's see if this helps.

**Comment by tonyfast** [comment]
The CPython changes to tokenize effect 3.6 & 3.7.  The `IPython` tests are not checking on Python3.6dev. 
 Should there be tests for 3.6dev because of this use case?

**Comment by Carreau** [comment]
Likely, yes. That would be good. unless it make the tests take too long in
High case we can test only on one -dev

**Comment by tonyfast** [comment]
If we consider a simple statement like `a=999`.

In Python 3.6 with the old tokenize:

```
Version info sys.version_info(major=3, minor=6, micro=6, releaselevel='final', serial=0)
[TokenInfo(type=1 (NAME), string='a', start=(1, 0), end=(1, 1), line='a=999'),
 TokenInfo(type=53 (OP), string='=', start=(1, 1), end=(1, 2), line='a=999'),
 TokenInfo(type=2 (NUMBER), string='999', start=(1, 2), end=(1, 5), line='a=999'),
 TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')]
```
compared to in Python 3.7 on release candidate.

```
Version info sys.version_info(major=3, minor=7, micro=1, releaselevel='candidate', serial=1)
[TokenInfo(type=1 (NAME), string='a', start=(1, 0), end=(1, 1), line='a=999'),
 TokenInfo(type=53 (OP), string='=', start=(1, 1), end=(1, 2), line='a=999'),
 TokenInfo(type=2 (NUMBER), string='999', start=(1, 2), end=(1, 5), line='a=999'),
 TokenInfo(type=4 (NEWLINE), string='', start=(1, 5), end=(1, 6), line=''),
 TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')]
```

**Comment by tonyfast** [comment]
@Carreau it looks like the core tests are passing, but I am receiving a timeout error in the terminal test group .  Do you have an insight into this?

**Comment by tonyfast** [comment]
The errors at the moment are in `SystemAssign`.  The test case `=!` is trying to assign a system statement.

**Comment by tonyfast** [comment]
Woo!  The new and old tests are passing now.  These fixes resolve #11334 #11354 #11346 #11337 .

@Carreau @takluyver could you review this when you get a chance?

**Comment by Carreau** [comment]
Thanks ! Awesome work @tonyfast !

I'll try to review and merced this week-end.

File: .travis.yml [hunk 1]
```diff
@@ -34,6 +34,7 @@ after_success:
 
 matrix:
     include:
+        - { python: "3.6-dev", dist: xenial, sudo: true }
```
Comment (Carreau) [inline .travis.yml]: I'm just going to add some "ok", to follow where I am in the review for myself.

Comment (Carreau) [reply .travis.yml]: ok.

File: IPython/core/inputtransformer2.py [hunk 2]
```diff
@@ -234,6 +234,7 @@ def find(cls, tokens_by_line):
         for line in tokens_by_line:
             assign_ix = _find_assign_op(line)
             if (assign_ix is not None) \
+                    and not line[assign_ix].line.strip().startswith('=') \
```
Comment (Carreau) [inline IPython/core/inputtransformer2.py]: ok

File: IPython/core/inputtransformer2.py [hunk 3]
```diff
@@ -369,11 +370,15 @@ def transform(self, lines):
         end_line = find_end_of_continued_line(lines, start_line)
         line = assemble_continued_line(lines, (start_line, start_col), end_line)
 
-        if line[:2] in ESCAPE_DOUBLES:
+        if len(line) > 1 and line[:2] in ESCAPE_DOUBLES:
```
Comment (Carreau) [inline IPython/core/inputtransformer2.py]: ok

File: IPython/core/inputtransformer2.py [hunk 4]
```diff
@@ -369,11 +370,15 @@ def transform(self, lines):
...
             escape, content = line[:2], line[2:]
         else:
             escape, content = line[:1], line[1:]
-        call = tr[escape](content)
+
+        if escape in tr:
+            call = tr[escape](content)
+        else:
+            call = ''
```
Comment (Carreau) [inline IPython/core/inputtransformer2.py]: ok

File: IPython/core/inputtransformer2.py [hunk 5]
```diff
@@ -575,10 +583,28 @@ def check_complete(self, cell: str):
...
+            elif character.strip():
+                break
+            else:
+                continue
+
+        if ends_with_newline:
+            # Append an newline for consistent tokenization
+            # See https://bugs.python.org/issue33899
+            cell += '\n'
+
```
Comment (Carreau) [inline IPython/core/inputtransformer2.py]: Seem good. Looks like we should either ends with 2 \n or None.

File: IPython/core/inputtransformer2.py [hunk 6]
```diff
@@ -575,10 +583,28 @@ def check_complete(self, cell: str):
...
         lines = cell.splitlines(keepends=True)
-        if lines[-1][:-1].endswith('\\'):
+
+        if not lines:
+            return 'complete', None
+
+        if lines[-1].endswith('\\'):
```
Comment (Carreau) [inline IPython/core/inputtransformer2.py]: ok

File: IPython/core/tests/test_inputtransformer2.py [hunk 7]
```diff
@@ -212,6 +214,17 @@ def test_check_complete():
     nt.assert_equal(cc("a = '''\n   hi"), ('incomplete', 3))
     nt.assert_equal(cc("def a():\n x=1\n global x"), ('invalid', None))
     nt.assert_equal(cc("a \\ "), ('invalid', None))  # Nothing allowed after backslash
+    nt.assert_equal(cc("1\\\n+2"), ('complete', None))
+    nt.assert_equal(cc("1\\\n+2"), ('complete', None))
```
Comment (bartskowron) [inline IPython/core/tests/test_inputtransformer2.py]: redundant test

**Comment by Carreau** [comment]
Thanks all, let's try to get that in and have an IPython 7.1 by end of week.
That should match the imminent release of Python 3.7.1 that have the compatibility issues mentioned earlier.

**Pull request merged**